csrc/multidevice/utils.cpp

```diff
@@ -96,4 +99,58 @@ bool isResharding(Expr* expr) {
   return !haveDifferentSharding(tv_ref, tvs).empty();
 }
 
+namespace {
+
+void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(ref->getDeviceMesh());
+  }
+  scheduler_utils::parallelizeAllLike(ref, tvs, {ParallelType::DIDx});
+}
+
+void reshardBefore(Expr* expr, Fusion* fusion) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "multi-output expressions are not supported");
+  NVF_ERROR(
+      expr->outputs().at(0)->isA<TensorView>(),
+      "the expression's output is not a TensorView");
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+  std::unordered_set<TensorView*> inputs;
+  std::transform(
+      expr->inputs().begin(),
+      expr->inputs().end(),
+      std::inserter(inputs, inputs.end()),
+      [](Val* val) {
+        NVF_ERROR(
+            val->isA<TensorView>(),
+            "the expression's input is not a TensorView");
+        return val->as<TensorView>();
+      });
+  std::vector<TensorView*> new_inputs;
+  for (auto input : haveDifferentSharding(output, inputs)) {
+    // TODO: reuse cacheAfter?
+    // TODO: here we should add a mechanism to potentially reuse the inserted
+    // resharding accross all the consumer of the resharded tensor. This way we
+    // could avoid wasteful resharding set insertion.
+    TensorView* new_input = set(input);
+    expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+    new_inputs.push_back(new_input);
+  }
+  if (!new_inputs.empty()) {
+    shardAllLike(output, new_inputs);
+  }
+}
+
+} // namespace
+
+void insertReshardings(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  for (auto expr : exprs) {
+    if (!isLowerableToCommunication(expr)) {
+      reshardBefore(expr, fusion);
```

Comment:
Just want to clarify a couple of things here. It seems a bit strange to do resharding before any expression that doesn't directly correspond to communication ops. For example, certainly scalar, non-tensor ops should not need resharding, and it seems it's implicitly assumed as there's assertions in `resharedBefore`. 

It seems that `insertReshardings` would fail if such ops exist in a given fusion. Shouldn't we make it work anyway? Likely just ignoring non-relevant ops.

Reply:
> It seems a bit strange to do resharding before any expression that doesn't directly correspond to communication ops. 

Actually that is quite the opposite: the point of this function is precisely to  insert a set before an expr that doesn't directly correspond to a communication op.

Consider the example:
```
tv0 = makeContigTensor(1);
tv0->setDeviceMesh({0});
tv1 = abs(tv0)
tv0->setDeviceMesh({1});
fusion.addInput(tv0);
fusion.addOutput(tv1);
```

Here, the unary op `abs` doesn't correspond directly to a communication op, because it embedds both a compute (taking the absolute value) AND a communication (send/recv from device 0 to device 1)

In this case, `insertReshardings` will add a intermediate tensor between `tv0` and `tv1`, i.e., it will transform the Fusion to be equivalent to:
```
tv0 = makeContigTensor(1);
tv0->setDeviceMesh({0});
tv0_ = set(tv0);
tv0_->setDeviceMesh({1});
tv1 = abs(tv0_)
tv0->setDeviceMesh({1});
fusion.addInput(tv0);
fusion.addOutput(tv1);
```

This way, we have "separated" the communication and the compute:
- `tv0_ = set(tv0)` corresponds directly to a communication between device 0 and 1
- `tv1 = abs(tv0_)` corresponds to a compute that is local to device 1, not involving any comm


> For example, certainly scalar, non-tensor ops should not need resharding, and it seems it's implicitly assumed as there's assertions in `resharedBefore`

Probably that some cases will need a special treatment, but we have not encoutered them so far. The case you are mentionning, that is "scalar, non-tensor ops", is a case we are not considering because we assert that all Vals are tensors. I think we should not bother with those cases until a concrete example comes to us.
Anyway, if we want (in this PR or later), to exclude some cases, we just need to add conditions on the line you mention. Let me know what condition(s) you would like me to add there.

Reply:
My comment is mostly about clarification rather than technical concerns.

From the function name, it seems to me that `isLowerableToCommunication` implies the given expr can be converted to communication ops, so it makes sense to say that if it returns `true`, it shouldn't need to reshard inputs. However, the fact that it returns `false` doesn't seem to me it requires sharding. A simple example is scalar ops, as I mentioned. 

So, to me more explicit, I think it would be more clear if the code would look like:

```
if (requireSharding(expr) && !isLowerableToCommunication(expr)) {
  reshardBefore(expr);
}
```

Does it make sense?

> Probably that some cases will need a special treatment, but we have not encoutered them so far.

Aren't we going to apply this routine to any given fusions in the future? It makes sense to have some assumptions for short term progress, but it'd be helpful to make it clear what such short-term assumptions are and make them clearly separated from more fundamental limitations.

Reply:
> My comment is mostly about clarification rather than technical concerns.

alright no problem

> From the function name, it seems to me that `isLowerableToCommunication` implies the given expr can be converted to communication ops, so it makes sense to say that if it returns `true`, it shouldn't need to have a separate reshard op for the inputs.

I agree

> However, the fact that it returns `false` doesn't seem to me it requires sharding. A simple example is scalar ops, as I mentioned.
> 
> So, to me more explicit, I think it would be more clear if the code would look like:
> 
> ```
> if (requireSharding(expr) && !isLowerableToCommunication(expr)) {
>   reshardBefore(expr);
> }
> ```
> 
> Does it make sense?

Ok I understand. Actually, there is the util function `isResharding` that can exactly fulfill this. The only reason I removed it is because, internally, `isResharding` calls `haveSameSharding` and checks if the returned set is empty. Since `haveSameSharding` is called anyway inside `reshardBefore`, this is wasteful.

So I did that for performance, over clarity. If you think we'd rather add `isResharding` for clarity let me know. Alternatively, I can also add a comment.

> > Probably that some cases will need a special treatment, but we have not encoutered them so far.
> 
> Aren't we going to apply this routine to any given fusions in the future? It makes sense to have some assumptions for short term progress, but it'd be helpful to make it clear what such short-term assumptions are and make them clearly separated from more fundamental limitations.

The routine proposed here is as general as I can imagine for now. The only assumption, made for simplicity, is that the Vals are TensorViews, and I think it is a reasonable assumption.

Let me know what do to with that

Reply:
I added a comment for clarity in the last commit. I resolve this conversation but feel free to open it again if needed

Reply:
> The routine proposed here is as general as I can imagine for now. The only assumption, made for simplicity, is that the Vals are TensorViews, and I think it is a reasonable assumption.

The assumption is fine, though I'm not sure why. The problem I had when reviewing the code, and still not clear, is that I didn't know there's this assumption. I added a comment to improve the clarity.

Reply:
> Ok I understand. Actually, there is the util function isResharding that can exactly fulfill this. The only reason I removed it is because, internally, isResharding calls haveSameSharding and checks if the returned set is empty. Since haveSameSharding is called anyway inside reshardBefore, this is wasteful.
> 
> So I did that for performance, over clarity. If you think we'd rather add isResharding for clarity let me know. Alternatively, I can also add a comment.

Again, my comment is for improving the code clarity by separating semantic concerns more clearly. What I'd expect the code should look like is:

```
for each expr in fusion:
  if needResharding(expr):
     insertReshardingBefore(expr)
```

`insertReshardings` exactly looks like this, but what's confusing is `reshardBefore` actually does additional conditional check and may not actually does resharding, and I don't understand why it needs to be written in that way. It's just clearing separating semantic concerns. The first thing is to check the condition, and the next thing is to do what needs to be done if the condition is met. 

I don't understand your performance concern. Of course, we don't want to call `haveSameSharding` twice. We just need to restructure the code, not adding additional things to do.

Reply:
> but what's confusing is reshardBefore actually does additional conditional check and may not actually does resharding

I renamed the function `reshardBefore` to `maybeReshardInputs`, perhaps this will make the code more readable.


> The first thing is to check the condition, and the next thing is to do what needs to be done if the condition is met.
I don't understand your performance concern. Of course, we don't want to call haveSameSharding twice.

On the one hand, the "condition" whether a "set insertion" is needed must be checked separately at each input -- on the other hand, `parallelizeAllLike` is more efficient if called on all the inputs at once.

That is why it is not so natural to divide it into two functions. Please look at the last version and let me know if it is better now.

Reply:
Thanks. Looks good now.